internal/xds/cache/snapshotcache.go

```diff
@@ -0,0 +1,271 @@
+// Portions of this code are based on code from Contour, available at:
+// https://github.com/projectcontour/contour/blob/main/internal/xds/v3/snapshotter.go
+
+package cache
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+
+	envoy_service_discovery_v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	envoy_cache_v3 "github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	envoy_server_v3 "github.com/envoyproxy/go-control-plane/pkg/server/v3"
+
+	"github.com/envoyproxy/gateway/internal/log"
+	"github.com/envoyproxy/gateway/internal/xds/types"
+)
+
+var Hash = envoy_cache_v3.IDHash{}
+
+// SnapshotCacheWithCallbacks uses the go-control-plane SimpleCache to store snapshots of
+// Envoy resources, sliced by Node ID so that we can do incremental xDS properly.
+// It does this by also implementing callbacks to make sure that the cache is kept
+// up to date for each new node.
+//
+// Having the cache also implement the callbacks is a little bit hacky, but it makes sure
+// that all the required bookkeeping happens.
+// TODO(youngnick): Talk to the go-control-plane maintainers and see if we can upstream
+// this in a better way.
+type SnapshotCacheWithCallbacks interface {
+	envoy_cache_v3.SnapshotCache
+	envoy_server_v3.Callbacks
+	GenerateNewSnapshot(types.XdsResources) error
+}
+
+type snapshotcache struct {
+	envoy_cache_v3.SnapshotCache
+
+	lastSnapshot *envoy_cache_v3.Snapshot
+
+	log *log.LogrWrapper
+
+	streamIDNodeID map[int64]string
+
+	snapshotVersion int64
+}
+
+// GenerateNewSnapshot takes a table of resources (the output from the IR->xDS
+// translator) and updates the snapshot version.
+func (s *snapshotcache) GenerateNewSnapshot(resources types.XdsResources) error {
+
+	version := s.newSnapshotVersion()
+
+	// Create a snapshot with all xDS resources.
+	snapshot, err := envoy_cache_v3.NewSnapshot(
+		version,
+		resources,
+	)
+	if err != nil {
+		return err
+	}
+
+	s.lastSnapshot = snapshot
+
+	for _, node := range s.getNodeIDs() {
+		fmt.Printf("Generating a snapshot with Node %s\n", node)
```

Comment:
lets make this a debug log

Reply:
Agreed, this should be a debug log, missed it, thanks.